File: SV.003-main/frontend/src/lib/clinicalChartSync.js
/**
 * Sync between species form_data and patients.clinical_chart.
 */

export const EMPTY_CLINICAL_CHART = {
  allergies: [],
  chronic_conditions: [],
  surgeries: [],
  vaccines: [],
  deworming: [],
  problems: [],
  reproductive: { sterilized: null, notes: "" },
  form_snapshot: null,
  form_category: null,
  updated_at: null,
  updated_from_consultation_id: null,
};

const LIST_KEYS = [
  "allergies",
  "chronic_conditions",
  "surgeries",
  "vaccines",
  "deworming",
  "problems",
];

function nowIso() {
  return new Date().toISOString();
}

function todayDate() {
  return nowIso().slice(0, 10);
}

export function normalizeClinicalChart(raw) {
  const base = {
    ...EMPTY_CLINICAL_CHART,
    reproductive: { ...EMPTY_CLINICAL_CHART.reproductive },
  };
  if (!raw || typeof raw !== "object") return base;
  const out = { ...base, ...raw };
  for (const key of LIST_KEYS) {
    out[key] = Array.isArray(out[key]) ? out[key] : [];
  }
  out.reproductive =
    out.reproductive && typeof out.reproductive === "object"
      ? { ...EMPTY_CLINICAL_CHART.reproductive, ...out.reproductive }
      : { ...EMPTY_CLINICAL_CHART.reproductive };
  out.form_snapshot =
    out.form_snapshot && typeof out.form_snapshot === "object" ? out.form_snapshot : null;
  out.form_category = out.form_category ? String(out.form_category) : null;
  return out;
}

function truthySi(value) {
  if (value === true) return true;
  const s = String(value || "")
    .trim()
    .toUpperCase();
  return ["SI", "SÍ", "YES", "TRUE", "1"].includes(s);
}

export function parseWeightKg(value) {
  if (value == null || value === "") return null;
  if (typeof value === "number" && Number.isFinite(value)) return value;
  const text = String(value).trim().replace(",", ".");
  const match = text.match(/(\d+(?:\.\d+)?)/);
  if (!match) return null;
  const n = Number(match[1]);
  return Number.isFinite(n) ? n : null;
}

function itemKey(item, fields) {
  return fields.map((f) => String(item?.[f] || "").trim().toLowerCase()).join("|");
}

function mergeList(existing, incoming, keyFields) {
  const merged = (existing || []).filter((x) => x && typeof x === "object").map((x) => ({ ...x }));
  const seen = new Set(merged.map((x) => itemKey(x, keyFields)));
  for (const item of incoming || []) {
    if (!item || typeof item !== "object") continue;
    if (!keyFields.some((f) => String(item[f] || "").trim())) continue;
    const key = itemKey(item, keyFields);
    if (seen.has(key)) continue;
    seen.add(key);
    merged.push({ ...item });
  }
  return merged;
}

export function mergeClinicalChart(existing, patch) {
  const base = normalizeClinicalChart(existing);
  if (!patch || typeof patch !== "object") return base;

  const out = { ...base };
  out.allergies = mergeList(base.allergies, patch.allergies, ["label", "noted_at"]);
  out.chronic_conditions = mergeList(base.chronic_conditions, patch.chronic_conditions, [
    "label",
    "noted_at",
  ]);
  out.surgeries = mergeList(base.surgeries, patch.surgeries, ["label", "date"]);
  out.vaccines = mergeList(base.vaccines, patch.vaccines, ["label", "date"]);
  out.deworming = mergeList(base.deworming, patch.deworming, ["type", "product", "date"]);

  if (Array.isArray(patch.problems)) {
    const byId = new Map();
    const order = [];
    for (const p of out.problems) {
      if (!p?.id) continue;
      byId.set(String(p.id), { ...p });
      order.push(String(p.id));
    }
    const openTitles = new Set(
      [...byId.values()]
        .filter((p) => (p.status || "open") !== "closed")
        .map((p) => String(p.title || "").trim().toLowerCase())
        .filter(Boolean),
    );
    for (const p of patch.problems) {
      if (!p || typeof p !== "object") continue;
      const titleKey = String(p.title || "").trim().toLowerCase();
      if (titleKey && openTitles.has(titleKey) && (p.status || "open") !== "closed") {
        continue;
      }
      const id = String(p.id || cryptoRandomId());
      if (byId.has(id)) {
        byId.set(id, { ...byId.get(id), ...p, id });
      } else {
        byId.set(id, { ...p, id });
        order.push(id);
        if (titleKey) openTitles.add(titleKey);
      }
    }
    out.problems = order.map((id) => byId.get(id)).filter(Boolean);
  }

  if (patch.reproductive && typeof patch.reproductive === "object") {
    out.reproductive = { ...out.reproductive, ...patch.reproductive };
  }
  if (patch.form_snapshot && typeof patch.form_snapshot === "object") {
    out.form_snapshot = { ...patch.form_snapshot };
  }
  if (patch.form_category != null && String(patch.form_category).trim()) {
    out.form_category = String(patch.form_category).trim();
  }
  if (patch.updated_from_consultation_id) {
    out.updated_from_consultation_id = patch.updated_from_consultation_id;
  }
  out.updated_at = patch.updated_at || nowIso();
  return out;
}

function cryptoRandomId() {
  if (typeof crypto !== "undefined" && crypto.randomUUID) return crypto.randomUUID();
  return `p_${Date.now()}_${Math.random().toString(36).slice(2, 9)}`;
}

/**
 * Extract a clinical_chart patch from species form fields.
 * Pass consultationId only from Diagnóstico (creates open problems from motivo).
 */
export function extractChartPatchFromFormData(category, formData, { consultationId, notedAt } = {}) {
  const fd = formData && typeof formData === "object" ? formData : {};
  const when = notedAt || todayDate();
  const patch = {
    vaccines: [],
    surgeries: [],
    deworming: [],
    chronic_conditions: [],
    allergies: [],
    problems: [],
    reproductive: {},
    updated_from_consultation_id: consultationId || null,
    updated_at: nowIso(),
  };

  if (truthySi(fd.vacunas_vigentes) || fd.vacunas_cual) {
    patch.vaccines.push({
      label: String(fd.vacunas_cual || "").trim() || "Vacunas vigentes",
      date: when,
      notes: "",
    });
  }

  for (const [key, label] of [
    ["vacuna_mixomatosis", "Mixomatosis"],
    ["vacuna_vhd", "VHD"],
    ["vacuna_rabia", "Rabia"],
  ]) {
    const raw = fd[key];
    if (!raw || String(raw).trim().toUpperCase() === "NO") continue;
    if (truthySi(raw) || String(raw).trim()) {
      const detail = String(fd[`${key}_fecha`] || fd[`${key}_cual`] || "").trim();
      patch.vaccines.push({ label, date: detail || when, notes: "" });
    }
  }

  if (truthySi(fd.cirugias_previas) || fd.cirugias_cual) {
    patch.surgeries.push({
      label: String(fd.cirugias_cual || "").trim() || "Cirugía previa",
      date: when,
      notes: "",
    });
  }

  if (truthySi(fd.desparasitacion_interna)) {
    const product = String(
      fd.desparasitacion_interna_cual || fd.desparasitacion_interna_producto || "",
    ).trim();
    const date = String(fd.desparasitacion_interna_fecha || when).trim();
    patch.deworming.push({ type: "internal", product: product || "Interna", date });
  }

  if (truthySi(fd.desparasitacion_externa)) {
    const product = String(
      fd.desparasitacion_externa_producto || fd.desparasitacion_externa_cual || "",
    ).trim();
    const date = String(fd.desparasitacion_externa_fecha || when).trim();
    patch.deworming.push({ type: "external", product: product || "Externa", date });
  }

  for (const [field, target] of [
    ["alergias", "allergies"],
    ["alergia", "allergies"],
    ["enfermedades_cronicas", "chronic_conditions"],
    ["enfermedades_previas", "chronic_conditions"],
  ]) {
    const raw = String(fd[field] || "").trim();
    if (!raw || ["NO", "N/A", "NA", "-"].includes(raw.toUpperCase())) continue;
    if (target === "allergies") {
      patch.allergies.push({ label: raw, severity: "", noted_at: when });
    } else {
      patch.chronic_conditions.push({ label: raw, status: "active", noted_at: when });
    }
  }

  const esterilizado = fd.esterilizado || fd.estado_reproductivo;
  if (esterilizado != null && String(esterilizado).trim() !== "") {
    const s = String(esterilizado).trim().toUpperCase();
    if (["SI", "SÍ", "YES", "ESTERILIZADO", "CASTRADO", "OVH", "ORQUIECTOMIA"].includes(s)) {
      patch.reproductive.sterilized = true;
    } else if (["NO", "ENTERO", "ENTERA", "NO ESTERILIZADO"].includes(s)) {
      patch.reproductive.sterilized = false;
    } else {
      patch.reproductive.notes = String(esterilizado).trim();
    }
  }

  const motivo = String(fd.motivo_consulta || fd.sintomas || "").trim();
  if (motivo && consultationId) {
    patch.problems.push({
      id: cryptoRandomId(),
      title: motivo.slice(0, 200),
      status: "open",
      opened_at: when,
      closed_at: "",
      source_consultation_id: consultationId,
    });
  }

  void category;
  return patch;
}

/**
 * Patch for saving species form as patient chart (no CDS consultation, no problems).
 */
export function buildSpeciesFormChartPatch(category, formData) {
  const patch = extractChartPatchFromFormData(category, formData, { consultationId: null });
  patch.problems = [];
  patch.updated_from_consultation_id = null;
  patch.form_snapshot = formData && typeof formData === "object" ? { ...formData } : {};
  patch.form_category = category || null;
  return patch;
}

function setIfEmpty(target, key, value) {
  if (value == null || value === "") return;
  if (target[key] == null || String(target[key]).trim() === "") {
    target[key] = value;
  }
}

/**
 * Prefill form fields from patient + clinical chart + last consultation.
 * Priority: form_snapshot > chart lists > patient demographics > last consultation.
 */
export function hydrateFormDataFromChart(category, chart, patient, lastConsultation) {
  const c = normalizeClinicalChart(chart || patient?.clinical_chart);
  const snapshot =
    c.form_snapshot && typeof c.form_snapshot === "object" ? { ...c.form_snapshot } : {};
  const lastFd =
    lastConsultation?.form_data && typeof lastConsultation.form_data === "object"
      ? lastConsultation.form_data
      : {};
  const next = { ...snapshot };

  if (patient?.name) setIfEmpty(next, "nombre_mascota", patient.name);
  if (patient?.breed) setIfEmpty(next, "raza", patient.breed);
  if (patient?.sex) setIfEmpty(next, "sexo", patient.sex);
  if (patient?.weight_kg != null) setIfEmpty(next, "peso", String(patient.weight_kg));
  const owner = patient?.clients;
  if (owner?.name) setIfEmpty(next, "nombre_dueño", owner.name);

  const lastVaccine = [...(c.vaccines || [])].reverse().find((v) => v?.label);
  if (lastVaccine) {
    setIfEmpty(next, "vacunas_vigentes", "SI");
    setIfEmpty(next, "vacunas_cual", lastVaccine.label);
  }

  const lastSurgery = [...(c.surgeries || [])].reverse().find((s) => s?.label);
  if (lastSurgery) {
    setIfEmpty(next, "cirugias_previas", "SI");
    setIfEmpty(next, "cirugias_cual", lastSurgery.label);
  }

  const lastInternal = [...(c.deworming || [])]
    .reverse()
    .find((d) => d?.type === "internal");
  if (lastInternal) {
    setIfEmpty(next, "desparasitacion_interna", "SI");
    setIfEmpty(next, "desparasitacion_interna_cual", lastInternal.product || "");
    setIfEmpty(next, "desparasitacion_interna_producto", lastInternal.product || "");
    setIfEmpty(next, "desparasitacion_interna_fecha", lastInternal.date || "");
  }

  const lastExternal = [...(c.deworming || [])]
    .reverse()
    .find((d) => d?.type === "external");
  if (lastExternal) {
    setIfEmpty(next, "desparasitacion_externa", "SI");
    setIfEmpty(next, "desparasitacion_externa_producto", lastExternal.product || "");
    setIfEmpty(next, "desparasitacion_externa_fecha", lastExternal.date || "");
  }

  if (c.reproductive?.sterilized === true) {
    setIfEmpty(next, "esterilizado", "SI");
    setIfEmpty(next, "estado_reproductivo", "ESTERILIZADO");
  } else if (c.reproductive?.sterilized === false) {
    setIfEmpty(next, "esterilizado", "NO");
    setIfEmpty(next, "estado_reproductivo", "ENTERO");
  } else if (c.reproductive?.notes) {
    setIfEmpty(next, "estado_reproductivo", c.reproductive.notes);
  }

  const allergies = (c.allergies || []).map((a) => a.label).filter(Boolean);
  if (allergies.length) setIfEmpty(next, "alergias", allergies.join("; "));

  const chronic = (c.chronic_conditions || [])
    .filter((x) => x.status !== "resolved")
    .map((x) => x.label)
    .filter(Boolean);
  if (chronic.length) setIfEmpty(next, "enfermedades_cronicas", chronic.join("; "));

  for (const key of [
    "habitat",
    "dieta",
    "zona_geografica",
    "alimentacion_seco",
    "alimentacion_humedo",
    "alimentacion_casero",
    "alimentacion_frecuencia",
  ]) {
    if (lastFd[key]) setIfEmpty(next, key, lastFd[key]);
  }

  void category;
  return next;
}

export function resolveChartFormCategory(chart, patient) {
  const c = normalizeClinicalChart(chart || patient?.clinical_chart);
  return c.form_category || patient?.species || "";
}

export function openProblems(chart) {
  const c = normalizeClinicalChart(chart);
  return (c.problems || []).filter((p) => p && p.status !== "closed");
}

export function latestByDate(items, dateField = "date") {
  if (!Array.isArray(items) || !items.length) return null;
  return [...items].sort((a, b) => String(b?.[dateField] || "").localeCompare(String(a?.[dateField] || "")))[0];
}

/** Categories supported by LazySpeciesForm (primary keys). */
export const SPECIES_FORM_CATEGORIES = [
  "perros",
  "gatos",
  "conejos",
  "aves",
  "hamsters",
  "cuyos",
  "hurones",
  "erizos",
  "tortugas",
  "iguanas",
  "patos_pollos",
];

File: SV.003-main/backend/clinical_chart_sync.py
"""Sync between species form_data and patients.clinical_chart."""

from __future__ import annotations

import re
import uuid
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Tuple

EMPTY_CHART: Dict[str, Any] = {
    "allergies": [],
    "chronic_conditions": [],
    "surgeries": [],
    "vaccines": [],
    "deworming": [],
    "problems": [],
    "reproductive": {"sterilized": None, "notes": ""},
    "form_snapshot": None,
    "form_category": None,
    "updated_at": None,
    "updated_from_consultation_id": None,
}

LIST_KEYS = (
    "allergies",
    "chronic_conditions",
    "surgeries",
    "vaccines",
    "deworming",
    "problems",
)


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def normalize_chart(raw: Any) -> Dict[str, Any]:
    base = {**EMPTY_CHART, "reproductive": {**EMPTY_CHART["reproductive"]}}
    if not isinstance(raw, dict):
        return base
    out = {**base, **raw}
    for key in LIST_KEYS:
        val = out.get(key)
        out[key] = list(val) if isinstance(val, list) else []
    repro = out.get("reproductive")
    out["reproductive"] = (
        {**EMPTY_CHART["reproductive"], **repro} if isinstance(repro, dict) else {**EMPTY_CHART["reproductive"]}
    )
    snap = out.get("form_snapshot")
    out["form_snapshot"] = snap if isinstance(snap, dict) else None
    cat = out.get("form_category")
    out["form_category"] = str(cat).strip() if cat else None
    return out


def _truthy_si(value: Any) -> bool:
    if value is True:
        return True
    s = str(value or "").strip().upper()
    return s in {"SI", "SÍ", "YES", "TRUE", "1"}


def _parse_weight(value: Any) -> Optional[float]:
    if value is None or value == "":
        return None
    if isinstance(value, (int, float)):
        return float(value)
    text = str(value).strip().replace(",", ".")
    match = re.search(r"(\d+(?:\.\d+)?)", text)
    if not match:
        return None
    try:
        return float(match.group(1))
    except ValueError:
        return None


def _item_key(item: Dict[str, Any], fields: Tuple[str, ...]) -> str:
    parts = [str(item.get(f) or "").strip().lower() for f in fields]
    return "|".join(parts)


def _merge_list(
    existing: List[Dict[str, Any]],
    incoming: List[Dict[str, Any]],
    key_fields: Tuple[str, ...],
) -> List[Dict[str, Any]]:
    merged = [dict(x) for x in existing if isinstance(x, dict)]
    seen = {_item_key(x, key_fields) for x in merged}
    for item in incoming:
        if not isinstance(item, dict):
            continue
        key = _item_key(item, key_fields)
        if not key or key == "|" * (len(key_fields) - 1) or key in seen:
            # allow empty-label skip
            if not any(str(item.get(f) or "").strip() for f in key_fields):
                continue
            if key in seen:
                continue
        seen.add(key)
        merged.append(dict(item))
    return merged


def merge_clinical_chart(existing: Any, patch: Any) -> Dict[str, Any]:
    base = normalize_chart(existing)
    if not isinstance(patch, dict):
        return base

    out = {**base}
    out["allergies"] = _merge_list(base["allergies"], patch.get("allergies") or [], ("label", "noted_at"))
    out["chronic_conditions"] = _merge_list(
        base["chronic_conditions"], patch.get("chronic_conditions") or [], ("label", "noted_at")
    )
    out["surgeries"] = _merge_list(base["surgeries"], patch.get("surgeries") or [], ("label", "date"))
    out["vaccines"] = _merge_list(base["vaccines"], patch.get("vaccines") or [], ("label", "date"))
    out["deworming"] = _merge_list(
        base["deworming"], patch.get("deworming") or [], ("type", "product", "date")
    )

    if isinstance(patch.get("problems"), list):
        # problems: replace-aware merge by id, else by title+opened_at
        by_id = {str(p.get("id")): dict(p) for p in out["problems"] if isinstance(p, dict) and p.get("id")}
        order = [str(p.get("id")) for p in out["problems"] if isinstance(p, dict) and p.get("id")]
        open_titles = {
            str(p.get("title") or "").strip().lower()
            for p in by_id.values()
            if (p.get("status") or "open") != "closed"
        }
        for p in patch["problems"]:
            if not isinstance(p, dict):
                continue
            title_key = str(p.get("title") or "").strip().lower()
            if title_key and title_key in open_titles and (p.get("status") or "open") != "closed":
                continue
            pid = str(p.get("id") or "") or str(uuid.uuid4())
            if pid in by_id:
                by_id[pid] = {**by_id[pid], **p, "id": pid}
            else:
                by_id[pid] = {**p, "id": pid}
                order.append(pid)
                if title_key:
                    open_titles.add(title_key)
        out["problems"] = [by_id[i] for i in order if i in by_id]

    if isinstance(patch.get("reproductive"), dict):
        out["reproductive"] = {**out["reproductive"], **patch["reproductive"]}

    if isinstance(patch.get("form_snapshot"), dict):
        out["form_snapshot"] = dict(patch["form_snapshot"])
    if patch.get("form_category") is not None and str(patch.get("form_category") or "").strip():
        out["form_category"] = str(patch["form_category"]).strip()

    if patch.get("updated_from_consultation_id"):
        out["updated_from_consultation_id"] = patch["updated_from_consultation_id"]
    out["updated_at"] = patch.get("updated_at") or _now_iso()
    return out


def extract_chart_patch_from_form_data(
    category: Optional[str],
    form_data: Optional[Dict[str, Any]],
    *,
    consultation_id: Optional[str] = None,
    noted_at: Optional[str] = None,
) -> Dict[str, Any]:
    fd = form_data if isinstance(form_data, dict) else {}
    when = noted_at or _now_iso()[:10]
    patch: Dict[str, Any] = {
        "vaccines": [],
        "surgeries": [],
        "deworming": [],
        "chronic_conditions": [],
        "allergies": [],
        "problems": [],
        "reproductive": {},
        "updated_from_consultation_id": consultation_id,
        "updated_at": _now_iso(),
    }

    # Vaccines (dogs/cats and similar)
    if _truthy_si(fd.get("vacunas_vigentes")) or fd.get("vacunas_cual"):
        label = str(fd.get("vacunas_cual") or "").strip() or "Vacunas vigentes"
        patch["vaccines"].append({"label": label, "date": when, "notes": ""})

    # Rabbit-specific vaccines
    for key, label in (
        ("vacuna_mixomatosis", "Mixomatosis"),
        ("vacuna_vhd", "VHD"),
        ("vacuna_rabia", "Rabia"),
    ):
        if _truthy_si(fd.get(key)) or (fd.get(key) and str(fd.get(key)).upper() not in {"NO", "FALSE", "0", ""}):
            if str(fd.get(key) or "").strip().upper() in {"NO", "FALSE", "0"}:
                continue
            detail = str(fd.get(f"{key}_fecha") or fd.get(f"{key}_cual") or "").strip()
            patch["vaccines"].append({"label": label, "date": detail or when, "notes": ""})

    # Surgeries
    if _truthy_si(fd.get("cirugias_previas")) or fd.get("cirugias_cual"):
        label = str(fd.get("cirugias_cual") or "").strip() or "Cirugía previa"
        patch["surgeries"].append({"label": label, "date": when, "notes": ""})

    # Deworming
    if _truthy_si(fd.get("desparasitacion_interna")):
        product = str(
            fd.get("desparasitacion_interna_cual")
            or fd.get("desparasitacion_interna_producto")
            or ""
        ).strip()
        date = str(fd.get("desparasitacion_interna_fecha") or when).strip()
        patch["deworming"].append({"type": "internal", "product": product or "Interna", "date": date})

    if _truthy_si(fd.get("desparasitacion_externa")):
        product = str(
            fd.get("desparasitacion_externa_producto")
            or fd.get("desparasitacion_externa_cual")
            or ""
        ).strip()
        date = str(fd.get("desparasitacion_externa_fecha") or when).strip()
        patch["deworming"].append({"type": "external", "product": product or "Externa", "date": date})

    # Allergies / chronic (free text fields when present)
    for field, target in (
        ("alergias", "allergies"),
        ("alergia", "allergies"),
        ("enfermedades_cronicas", "chronic_conditions"),
        ("enfermedades_previas", "chronic_conditions"),
    ):
        raw = str(fd.get(field) or "").strip()
        if raw and raw.upper() not in {"NO", "N/A", "NA", "-"}:
            item = {"label": raw, "noted_at": when}
            if target == "allergies":
                item["severity"] = ""
                patch["allergies"].append(item)
            else:
                item["status"] = "active"
                patch["chronic_conditions"].append(item)

    # Reproductive
    esterilizado = fd.get("esterilizado") or fd.get("estado_reproductivo")
    if esterilizado is not None and str(esterilizado).strip() != "":
        s = str(esterilizado).strip().upper()
        if s in {"SI", "SÍ", "YES", "ESTERILIZADO", "CASTRADO", "OVH", "ORQUIECTOMIA"}:
            patch["reproductive"]["sterilized"] = True
        elif s in {"NO", "ENTERO", "ENTERA", "NO ESTERILIZADO"}:
            patch["reproductive"]["sterilized"] = False
        else:
            patch["reproductive"]["notes"] = str(esterilizado).strip()

    # Open problem from motivo
    motivo = str(fd.get("motivo_consulta") or fd.get("sintomas") or "").strip()
    if motivo and consultation_id:
        patch["problems"].append(
            {
                "id": str(uuid.uuid4()),
                "title": motivo[:200],
                "status": "open",
                "opened_at": when,
                "closed_at": "",
                "source_consultation_id": consultation_id,
            }
        )

    # Drop empty lists so merge can skip noise when replacing via API full lists
    return patch


def extract_weight_kg(form_data: Optional[Dict[str, Any]]) -> Optional[float]:
    if not isinstance(form_data, dict):
        return None
    return _parse_weight(form_data.get("peso"))


def build_vitals_series(consultations: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Build chronological vitals from serialized or raw consultation rows."""
    series: List[Dict[str, Any]] = []
    for row in consultations or []:
        if not isinstance(row, dict):
            continue
        # Accept already-serialized or raw DB row
        payload = row.get("payload") if isinstance(row.get("payload"), dict) else {}
        form_data = row.get("form_data")
        if not isinstance(form_data, dict):
            form_data = (payload or {}).get("form_data") or (payload or {}).get("consultation_data") or {}
        parametros = row.get("parametros_vitales")
        if parametros is None:
            parametros = (payload or {}).get("parametros_vitales")
        weight = _parse_weight(form_data.get("peso") if isinstance(form_data, dict) else None)
        if weight is None and isinstance(parametros, dict):
            weight = _parse_weight(parametros.get("peso") or parametros.get("weight") or parametros.get("weight_kg"))
        entry = {
            "consultation_id": row.get("id"),
            "created_at": row.get("created_at"),
            "weight_kg": weight,
            "parametros_vitales": parametros if isinstance(parametros, dict) else None,
        }
        if entry["weight_kg"] is not None or entry["parametros_vitales"]:
            series.append(entry)
    series.sort(key=lambda x: x.get("created_at") or "")
    return series
